### PR TITLE
Migrate parameters to v14.1 branch

### PR DESCRIPTION
**Pull request by caco3**
This PR adds a migration function of some parameters:

| **Section**     | **Old Name**           | **New Name**              | **Note**
|-----------------|------------------------|---------------------------|--|
| **MakeImage**   |    -        |    -            |  Renamed section from `MakeImage` to `TakeImage`|
| **TakeImage**   | `LogImageLocation`       | `RawImagesLocation`         |  Renamed Parameter |
|                 | `LogfileRetentionInDays` | `RawImagesRetention`        |  Renamed Parameter |
|                 | `Demo` |    -     | If disabled, set to default value, enable it |
|                 | `FixedExposure` |    -     | If disabled, set to default value, enable it |
|                 | `InitialMirror` |   -     | If disabled, set to default value, enable it |
|                 | `FlipImageSize` |    -     | If disabled, set to default value, enable it |
| **Digits**      | `LogImageLocation`       | `ROIImagesLocation`       |  Renamed Parameter |
|                 | `LogfileRetentionInDays` | `ROIImagesRetention`      |  Renamed Parameter |
| **Analog**      | `LogImageLocation`       | `ROIImagesLocation`      |  Renamed Parameter |
|                 | `LogfileRetentionInDays` | `ROIImagesRetention`     |  Renamed Parameter |
|                 | `ExtendedResolution` |      |  Removed as no longer used (use `<NUMBER>.ExtendedResolution` instead) |
| **PostProcessing** | `PreValueUse` |    -     | If disabled, set to default value, enable it |
|  | `<NUMBER>.ExtendedResolution` |    -     | If disabled, set to default value, enable it |
|  | `<NUMBER>.IgnoreLeadingNaN` |    -     | If disabled, set to default value, enable it |
|  | `AllowNegativeRates` |   `<NUMBER>.AllowNegativeRates`     | If disabled, set to default value, enable it<br>N.B. the `<NUMBER>` prefix was missing in the default config |
|  | `ErrorMessage` |  -       | If disabled, set to default value, enable it |
| **MQTT**      | `SetRetainFlag`         | `RetainMessages`             |  If disabled, set to default value, enable it |
|  | `Topic` | `MainTopic`        |  Renamed Parameter |
| **InfluxDB**   | `DataLogActive`                |        -        | If disabled, set to default value, enable it  |
| **DataLogging** | `DataLogRetentionInDays` | `DataFilesRetention`        |  Renamed Parameter |
| **AutoTimer**   | `Intervall`                | `Interval`                  |  Renamed Parameter |
|    | `AutoStart`                |       -         | If disabled, set to default value, enable it  |
| **Debug**   | `Logfile`                | `LogLevel`                  |  And switch form boolean `True`/`False` to `WARNING`|
|       | `LogfileRetentionInDays`         | `LogfilesRetention`             |  If disabled, set to default value, enable it |
| **System**      | `RSSIThreashold`         | `RSSIThreshold`             |  Renamed Parameter |
|                 | `AutoAdjustSummertime `  |  -|  Removed, no longer in use |

Also it switches all disabled boolean parameters to their default value . Then enable them.

The corresponding parts in the firmware and Web UI got also updated accordingly.

The migration function gets called at startup and modifies the config file as needed. If one or more parameter needs a migration, it stores the `config.ini` as `config.bak` **before** any change to it.
The migration has some simplifications:
 - Case Sensitiveness must be like in the initial config.ini
 - No Whitespace after a semicollon
 - Only one whitespace before/after the equal sign

And all changes are documented in the log file (as ERROR so they show up for sure):
```

[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line '[MakeImage]' to '[TakeImage]'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogImageLocation = /log/source' to ';CamImagesLocation = /log/source'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogfileRetentionInDays = 15' to ';CamImagesRetention = 15'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogImageLocation = /log/digit' to ';ROIImagesLocation = /log/digit'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogfileRetentionInDays = 3' to ';ROIImagesRetention = 3'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';[Analog]' to '[Analog]'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogImageLocation = /log/analog' to ';ROIImagesLocation = /log/analog'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';LogfileRetentionInDays = 3' to ';ROIImagesRetention = 3'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';SetRetainFlag = true' to ';RetainMessages = true'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';[InfluxDB]' to '[InfluxDB]'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';[GPIO]' to '[GPIO]'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line 'Intervall = 2' to 'Interval = 2'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line 'DataLogRetentionInDays = 3' to 'DataFilesRetention = 3'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line 'Logfile = 3' to 'LogLevel = 3'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line 'LogfileRetentionInDays = 30' to 'LogfilesRetention = 30'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Migrated Configfile line ';AutoAdjustSummertime = false' to ';;UNUSED_PARAMETER = false'
[0d00h00m04s] 2023-02-12T18:41:30 <ERR> [MAIN] Config file migrated. Saved backup to /sdcard/config/config.bak
```

See also discussions in:
- https://github.com/jomjol/AI-on-the-edge-device/pull/2017
- https://github.com/jomjol/AI-on-the-edge-device/pull/2001

### CODE REVIEW

**Comment by jomjol** [comment]
Started to test and found the following issue with removing the Enableling Checkbox for true/false values.
```
;main.IgnoreLeadingNaN = false
;ErrorMessage = true
```
In both cases the leading ";" should be removed. The second one is easy in the migration function:
`migrated = migrated | replace(configLines[i], ";ErrorMessage", "ErrorMessage");`

The first one is a problem, as the name of the number is not known!

We need to solve this.

**Comment by Slider0007** [comment]
@caco3: I did a quick test. Renaming works like described :-)

Some first comments / remarks:

- Saving CamImages, digit ROIImages and analog ROIImages not working anymore. Files are not stored when activating respective functions (I haven't checked retention).

- `Set Retain Flag` keeps greyed out if selected + text in UI needs to be adapted to new name `RetainMessages`
image

- Name `Logfile Log Level` differs from parameter name `LogLevel`, by purpose?

- `CamImages...` --> maybe `RawImages...`?
- `LogfilesRetention` / `DataFilesRetention` --> Align ...filesRetention or ...FilesRetention?

**Comment by Slider0007** [comment]
`HomeAssistant` keeps unchecked even after saving...

image

@caco3: I'm a little confused. Is the removing of the checkboxes for boolean parameters already included in this PR or will this be done with a separate PR?

**Comment by jomjol** [comment]
I have removed the checkboxes (again? - thought I already did it). Additionally I have updated the migrate funtion to remove the ";" for the simple parameters (without a number).
For the one with a number (e.g. HT.AllowNegativeRates = true) we still need to extend the migration function.

File: code/main/main.cpp [hunk 1]
```diff
@@ -474,7 +474,10 @@ void migrateConfiguration(void) {
         if (section == "[MQTT]") {
             migrated = migrated | replace(configLines[i], ";RetainMessages", "RetainMessages");
             migrated = migrated | replace(configLines[i], "SetRetainFlag", "RetainMessages");
-            migrated = migrated | replace(configLines[i], "MainTopic", "Topic");
+            if (configLines[i].rfind("Topic", 0) != std::string::npos)  // only if string starts with "Topic"
```
Comment (caco3) [inline code/main/main.cpp:477]: @jomjol Why is this needed? Isn't already before called `MainTopic`?

**Comment by caco3** [comment]
>     * `CamImages...` --> maybe `RawImages...`?

Done

>     * `LogfilesRetention` / `DataFilesRetention` --> Align ...filesRetention or ...FilesRetention?

I wanted to do as suggested, but then i learned that there is a common class for all them, so I wanted to keep it consistent.

**Comment by jomjol** [comment]
Is everything done and we can merge it?

**Comment by caco3** [comment]
Saving CamImages, digit ROIImages and analog ROIImages not working anymore. Files are not stored when activating respective functions (I haven't checked retention).

- [ ] I can confirm this, but do not understand the reason yet.

**Comment by caco3** [comment]
Saving CamImages, digit ROIImages and analog ROIImages is working now

**Comment by caco3** [comment]
@Slider0007 @jomjol I hope I have now everything. Please give it another try.

**Comment by jomjol** [comment]
> @Slider0007 @jomjol I hope I have now everything. Please give it another try.

I cannot test before the weekend.

**Comment by Slider0007** [comment]
@caco3: I did a quick test of UI config page.

- Extended resolution is greyed out
image

- Could the checkboxes of those parameters also get removed?
image
image

Does it make sense to add an indication that this config.ini is a migrated one or is it clear due to the new names? (config.ini - firmware version > 14.1).
I was just thinking about the fact when somebody goes back to an older version... (I tried it, system is behaving wierd)

I'll do further testing tomorrow...

**Comment by caco3** [comment]
>     * Extended resolution is greyed out
>       image

Oh bugger, somehow I missed that, thanks!

>     * Could the checkboxes of those parameters also get removed?
>       image
>       image

That question leads to the question if a parameter should be able to be disabled at all.
Personally I think the way the configuration is setup is highly overcomplicated. A parameter should either be set, or use its default if it is missing. Here we have some tri-state situation.
But I do not want to dig into it more and break it further.
My main goal was to rename them as needed so we can properly document them.
We can think of removing all those checkboxes for all parameters, but lets do it at a later time.

> Does it make sense to add an indication that this config.ini is a migrated one or is it clear due to the new names? (config.ini - firmware version > 14.1). I was just thinking about the fact when somebody goes back to an older version... (I tried it, system is behaving wierd)

Based on the naming it should be clear.
But I did not plan to provide downgrade support below 14.1.
Thats actually the main reason I want it in 14.1. 14.0 is quite stable as it looks, and in 14.x we will change a lot again. so people can go back to the 14.1 release without issues.
Notabene this will need to be state clearly in the release notes.
And like we already always say, a backup is highly recommended. Additionally before a migration it does a backup.
 
> I'll do further testing tomorrow...

Thank you!

**Comment by Slider0007** [comment]
> That question leads to the question if a parameter should be able to be disabled at all.
> Personally I think the way the configuration is setup is highly overcomplicated. A parameter should either be set, or use its default if it is missing. Here we have some tri-state situation.
> But I do not want to dig into it more and break it further.
> My main goal was to rename them as needed so we can properly document them.
> We can think of removing all those checkboxes for all parameters, but lets do it at a later time.

That's true, so it's totally fine for me 👍

**Comment by Slider0007** [comment]
@caco3: From my prespective everthing seems OK so far :-)

Only one minor: ClassFlowMakeImage.h and ClassFlowMakeImage.cpp are now empty in project. Can they be deleted?

**Comment by caco3** [comment]
> @caco3: From my prespective everthing seems OK so far :-)

Thanks!


> Only one minor: ClassFlowMakeImage.h and ClassFlowMakeImage.cpp are now empty in project. Can they be deleted?

Right, they got re-added by accident.


@jomjol I will merge now and rebase rolling to it so I can proceed with https://github.com/jomjol/AI-on-the-edge-device/pull/1986

If you find some issues, we can add another PR to the 14.1 branch

**Comment by Slider0007** [comment]
@caco3: Using the latest rolling I figured out two more things which needs to be adressed:

- Checkbox available, black text but unchecked
image

- Topic parameter in config.ini after saving from UI --> I thing this is related to this
https://github.com/jomjol/AI-on-the-edge-device/blob/23d2ae627d14843e5cc40ea0642c2905fdb2363e/sd-card/html/readconfigparam.js#L180
image

**Comment by caco3** [comment]
> AllowNegativeRates

hmm, kannst du das reproduzieren?

> ;Topic = 

Ah, now I understand.
If I understand @jomjol correctly, `Topic` was used a long time ago but is no longer used. not sure why you ended up to have this line

**Comment by jomjol** [comment]
It is still in the `config.ini` because due to downwards compatibility is was still as a parameter in the `readconfigparam.js` and every parameter in there get written back to the `config.ini` once it is saved again. Solved in the latest rolling
